### PR TITLE
feat: add track_timestamps_staleness limit to CRDs

### PR DESCRIPTION
**Pull request by dongjiang1989**
## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

ref: https://github.com/prometheus/prometheus/pull/13060

add track_timestamps_staleness limit to CRDs

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
add track_timestamps_staleness limit to CRDs
```

### CODE REVIEW

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 1]
```diff
@@ -423,6 +423,11 @@ type CommonPrometheusFields struct {
 	// Otherwise the HonorTimestamps field of the service or pod monitor applies.
 	OverrideHonorTimestamps bool `json:"overrideHonorTimestamps,omitempty"`
 
+	// When true, Prometheus ignores the timestamps for all the targets created
+	// from service and pod monitors.
+	// Otherwise the TrackTimestampsStaleness field of the service or pod monitor applies.
+	OverrideTrackTimestampsStaleness bool `json:"overrideTrackTimestampsStaleness,omitempty"`
```
Comment (nicolastakashi) [inline pkg/apis/monitoring/v1/prometheus_types.go:429]: Quick question.
Do we see any use cases where Prometheus admin wants to override the decision made by the scrape config?

**Comment by nicolastakashi** [comment]
Cool!
Thanks for handling that.

File: pkg/apis/monitoring/v1/prometheus_types.go (hunk 1 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1/prometheus_types.go:429]: When `OverrideTrackTimestampsStaleness` is set to "true" for Job target, it will be inserted in the TSDB when the target is down.

File: pkg/apis/monitoring/v1/podmonitor_types.go [hunk 2]
```diff
@@ -207,6 +207,11 @@ type PodMetricsEndpoint struct {
 	// +optional
 	HonorTimestamps *bool `json:"honorTimestamps,omitempty"`
 
+	// TrackTimestampsStaleness whether Prometheus tracks staleness of
+	// the metrics that have an explicit timestamps present in scraped data.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/podmonitor_types.go:211]:
```suggestion
	// the metrics that have an explicit timestamp present in scraped data.
```

File: pkg/apis/monitoring/v1/types.go [hunk 3]
```diff
@@ -438,6 +438,14 @@ type Endpoint struct {
 	// +optional
 	HonorTimestamps *bool `json:"honorTimestamps,omitempty"`
 
+	// TrackTimestampsStaleness whether Prometheus tracks staleness of
+	// the metrics that have an explicit timestamps present in scraped data.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:442]:
```suggestion
	// the metrics that have an explicit timestamp present in scraped data.
```

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 4]
```diff
@@ -135,6 +135,13 @@ type ScrapeConfigSpec struct {
 	// HonorTimestamps controls whether Prometheus respects the timestamps present in scraped data.
 	// +optional
 	HonorTimestamps *bool `json:"honorTimestamps,omitempty"`
+	// TrackTimestampsStaleness whether Prometheus tracks staleness of
+	// the metrics that have an explicit timestamps present in scraped data.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:139]:
```suggestion
	// the metrics that have an explicit timestamp present in scraped data.
```

Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:139]: Thanks for your review. 
Done

File: pkg/apis/monitoring/v1/types.go (hunk 3 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1/types.go:442]: Thanks for your review.
Done

File: pkg/apis/monitoring/v1/podmonitor_types.go (hunk 2 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1/podmonitor_types.go:211]: Thanks for your review.
Done

File: pkg/apis/monitoring/v1/prometheus_types.go (hunk 1 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1/prometheus_types.go:429]: @nicolastakashi @simonpasquier  Is it necessary to retain the `OverrideTrackTimestampsStaleness` global configuration?🤔

Comment (simonpasquier) [reply pkg/apis/monitoring/v1/prometheus_types.go:429]: I would keep it simple and drop it for now.

Comment (nicolastakashi) [reply pkg/apis/monitoring/v1/prometheus_types.go:429]: From my understanding we can remove it for now and focus only on svc/pod monitor as @simonpasquier mentioned.

File: jsonnet/prometheus-operator/prometheusagents-crd.json [hunk 5]
```diff
@@ -3652,6 +3652,10 @@
                     "description": "When true, Prometheus ignores the timestamps for all the targets created from service and pod monitors. Otherwise the HonorTimestamps field of the service or pod monitor applies.",
                     "type": "boolean"
                   },
+                  "overrideTrackTimestampsStaleness": {
+                    "description": "When true, Prometheus ignores the timestamps for all the targets created from service and pod monitors. Otherwise the TrackTimestampsStaleness field of the service or pod monitor applies.",
```
Comment (jan--f) [inline jsonnet/prometheus-operator/prometheusagents-crd.json:3656]: The description here needs and update.

File: jsonnet/prometheus-operator/prometheuses-crd.json [hunk 6]
```diff
@@ -4090,6 +4090,10 @@
                     "description": "When true, Prometheus ignores the timestamps for all the targets created from service and pod monitors. Otherwise the HonorTimestamps field of the service or pod monitor applies.",
                     "type": "boolean"
                   },
+                  "overrideTrackTimestampsStaleness": {
+                    "description": "When true, Prometheus ignores the timestamps for all the targets created from service and pod monitors. Otherwise the TrackTimestampsStaleness field of the service or pod monitor applies.",
```
Comment (jan--f) [inline jsonnet/prometheus-operator/prometheuses-crd.json:4094]: The description here needs and update.

File: pkg/apis/monitoring/v1/podmonitor_types.go (hunk 2 above)
Comment (jan--f) [inline pkg/apis/monitoring/v1/podmonitor_types.go:211]: Is it worth adding `Has no effect if no timestamps are exposed or `honorTimestamps` is false.`?

File: pkg/apis/monitoring/v1/prometheus_types.go (hunk 1 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1/prometheus_types.go]: Thanks @simonpasquier @nicolastakashi , PTAL
Drop the `OverrideTrackTimestampsStaleness` global configuration.

File: jsonnet/prometheus-operator/prometheusagents-crd.json (hunk 5 above)
Comment (dongjiang1989) [reply jsonnet/prometheus-operator/prometheusagents-crd.json]: Remove `overrideTrackTimestampsStaleness`.
Please re-check.

File: jsonnet/prometheus-operator/prometheuses-crd.json (hunk 6 above)
Comment (dongjiang1989) [reply jsonnet/prometheus-operator/prometheuses-crd.json]: Remove `overrideTrackTimestampsStaleness`.
Please re-check.

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 7]
```diff
@@ -422,7 +422,6 @@ type CommonPrometheusFields struct {
 	// from service and pod monitors.
 	// Otherwise the HonorTimestamps field of the service or pod monitor applies.
 	OverrideHonorTimestamps bool `json:"overrideHonorTimestamps,omitempty"`
-
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:425]: can you revert this change

File: pkg/apis/monitoring/v1/podmonitor_types.go [hunk 8]
```diff
@@ -207,6 +207,11 @@ type PodMetricsEndpoint struct {
 	// +optional
 	HonorTimestamps *bool `json:"honorTimestamps,omitempty"`
 
+	// TrackTimestampsStaleness whether Prometheus tracks staleness of
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/podmonitor_types.go:210]:
```suggestion
	// `trackTimestampsStaleness` defines whether Prometheus tracks staleness of
```

File: pkg/apis/monitoring/v1/types.go [hunk 9]
```diff
@@ -438,6 +438,14 @@ type Endpoint struct {
 	// +optional
 	HonorTimestamps *bool `json:"honorTimestamps,omitempty"`
 
+	// TrackTimestampsStaleness whether Prometheus tracks staleness of
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:441]:
```suggestion
	// `trackTimestampsStaleness` defines whether Prometheus tracks staleness of
```

File: pkg/apis/monitoring/v1/podmonitor_types.go [hunk 10]
```diff
@@ -207,6 +207,11 @@ type PodMetricsEndpoint struct {
...
+	// the metrics that have an explicit timestamps present in scraped data.
+	// It requires Prometheus >= v2.48.0.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/podmonitor_types.go:212]:
```suggestion
	//
	// It requires Prometheus >= v2.48.0.
	//
	// +optional
```

File: pkg/prometheus/promcfg.go [hunk 11]
```diff
@@ -247,6 +247,23 @@ func (cg *ConfigGenerator) AddHonorTimestamps(cfg yaml.MapSlice, userHonorTimest
 	return cg.WithMinimumVersion("2.9.0").AppendMapItem(cfg, "honor_timestamps", honor && !cpf.OverrideHonorTimestamps)
 }
 
+// AddTrackTimestampsStaleness adds the track_timestamps_staleness field into scrape configurations.
+// For backwards compatibility with Prometheus <2.48.0 we don't set
+// track_timestamps_staleness.
+func (cg *ConfigGenerator) AddTrackTimestampsStaleness(cfg yaml.MapSlice, userTrackTimestampsStaleness *bool) yaml.MapSlice {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:253]:
```suggestion
func (cg *ConfigGenerator) AddTrackTimestampsStaleness(cfg yaml.MapSlice, trackTimestampsStaleness *bool) yaml.MapSlice {
```

File: pkg/prometheus/promcfg.go [hunk 12]
```diff
@@ -247,6 +247,23 @@ func (cg *ConfigGenerator) AddHonorTimestamps(cfg yaml.MapSlice, userHonorTimest
...
+	if userTrackTimestampsStaleness == nil {
+		return cfg
+	}
+
+	track := false
+	if userTrackTimestampsStaleness != nil {
+		track = *userTrackTimestampsStaleness
+	}
+
+	return cg.WithMinimumVersion("2.48.0").AppendMapItem(cfg, "track_timestamps_staleness", track)
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:264]:
```suggestion
	
	return cg.WithMinimumVersion("2.48.0").AppendMapItem(cfg, "track_timestamps_staleness", trackTimestampsStaleness)
```

Comment (dongjiang1989) [reply pkg/prometheus/promcfg.go:264]: Done.

File: pkg/prometheus/promcfg.go (hunk 11 above)
Comment (dongjiang1989) [reply pkg/prometheus/promcfg.go:253]: Done

File: pkg/apis/monitoring/v1/prometheus_types.go (hunk 7 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1/prometheus_types.go]: Done

**Comment by dongjiang1989** [comment]
@simonpasquier addressed comments, please review :)

**Review by nicolastakashi — APPROVED** [review]

File: jsonnet/prometheus-operator/servicemonitors-crd.json [hunk 13]
```diff
@@ -626,6 +626,10 @@
                             }
                           },
                           "type": "object"
+                        },
+                        "trackTimestampsStaleness": {
+                          "description": "`trackTimestampsStaleness` defines whether Prometheus tracks staleness of the metrics that have an explicit timestamp present in scraped data. \n It requires Prometheus >= v2.48.0.",
```
Comment (jan--f) [inline jsonnet/prometheus-operator/servicemonitors-crd.json:631]:
```suggestion
                          "description": "`trackTimestampsStaleness` defines whether Prometheus tracks staleness of the metrics that have an explicit timestamp present in scraped data. Has no effect if `honorTimestamps` is false. It requires Prometheus >= v2.48.0.",
```

File: jsonnet/prometheus-operator/scrapeconfigs-crd.json [hunk 14]
```diff
@@ -1550,6 +1550,10 @@
                       }
                     },
                     "type": "object"
+                  },
+                  "trackTimestampsStaleness": {
+                    "description": "TrackTimestampsStaleness whether Prometheus tracks staleness of the metrics that have an explicit timestamp present in scraped data. Has no effect if `honorTimestamps` is false. It requires Prometheus >= v2.48.0.",
```
Comment (jan--f) [inline jsonnet/prometheus-operator/scrapeconfigs-crd.json:1555]:
```suggestion
                    "description": "`trackTimestampsStaleness` defines whether Prometheus tracks staleness of the metrics that have an explicit timestamp present in scraped data. Has no effect if `honorTimestamps` is false. It requires Prometheus >= v2.48.0.",
```

File: jsonnet/prometheus-operator/podmonitors-crd.json [hunk 15]
```diff
@@ -651,6 +651,10 @@
                             }
                           },
                           "type": "object"
+                        },
+                        "trackTimestampsStaleness": {
+                          "description": "TrackTimestampsStaleness whether Prometheus tracks staleness of the metrics that have an explicit timestamps present in scraped data. Has no effect if `honorTimestamps` is false. \n It requires Prometheus >= v2.48.0.",
```
Comment (jan--f) [inline jsonnet/prometheus-operator/podmonitors-crd.json:656]:
```suggestion
                          "description": "`trackTimestampsStaleness` defines whether Prometheus tracks staleness of the metrics that have an explicit timestamp present in scraped data. Has no effect if `honorTimestamps` is false. It requires Prometheus >= v2.48.0.",
```

File: pkg/apis/monitoring/v1/podmonitor_types.go [hunk 16]
```diff
@@ -207,6 +207,15 @@ type PodMetricsEndpoint struct {
 	// +optional
 	HonorTimestamps *bool `json:"honorTimestamps,omitempty"`
 
+	// TrackTimestampsStaleness whether Prometheus tracks staleness of
+	// the metrics that have an explicit timestamps present in scraped data.
```
Comment (jan--f) [inline pkg/apis/monitoring/v1/podmonitor_types.go:211]:
```suggestion
	// the metrics that have an explicit timestamp present in scraped data.
```

File: pkg/apis/monitoring/v1/podmonitor_types.go [hunk 17]
```diff
@@ -207,6 +207,15 @@ type PodMetricsEndpoint struct {
 	// +optional
 	HonorTimestamps *bool `json:"honorTimestamps,omitempty"`
 
+	// TrackTimestampsStaleness whether Prometheus tracks staleness of
```
Comment (jan--f) [inline pkg/apis/monitoring/v1/podmonitor_types.go:210]:
```suggestion
	// `trackTimestampsStaleness` defines whether Prometheus tracks staleness of
```

File: pkg/apis/monitoring/v1/types.go [hunk 18]
```diff
@@ -438,6 +438,14 @@ type Endpoint struct {
 	// +optional
 	HonorTimestamps *bool `json:"honorTimestamps,omitempty"`
 
+	// `trackTimestampsStaleness` defines whether Prometheus tracks staleness of
+	// the metrics that have an explicit timestamp present in scraped data.
```
Comment (jan--f) [inline pkg/apis/monitoring/v1/types.go:442]:
```suggestion
	// the metrics that have an explicit timestamp present in scraped data.
	// Has no effect if `honorTimestamps` is false.
```

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 19]
```diff
@@ -135,6 +135,13 @@ type ScrapeConfigSpec struct {
 	// HonorTimestamps controls whether Prometheus respects the timestamps present in scraped data.
 	// +optional
 	HonorTimestamps *bool `json:"honorTimestamps,omitempty"`
+	// TrackTimestampsStaleness whether Prometheus tracks staleness of
```
Comment (jan--f) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:138]:
```suggestion
	// `trackTimestampsStaleness` whether Prometheus tracks staleness of
```

**Review by jan--f — COMMENTED** [review]
Just a few nits for consistency in the doc comments. Otherwise this looks good!

**Review by simonpasquier — APPROVED** [review]
Thanks!

**Comment by simonpasquier** [comment]
I'm not dismissing the latest comments from @jan--f but they can be addressed in a follow-up PR :)
